documentation/dataplane/dash-flow-api.md

```diff
@@ -0,0 +1,633 @@
+| SAI_FLOW_ENTRY_BULK_GET_SESSION_ATTR_FIFTH_FLOW_ENTRY_BULK_GET_SESSION_FILTER_ID | @type: `sai_object_id_t` @objects `SAI_OBJECT_TYPE_FLOW_ENTRY_BULK_GET_SESSION_FILTER` | Action set_flow_entry_bulk_get_session_attr parameter THIRD_FLOW_ENTRY_BULK_GET_SESSION_FILTER_ID |
+
+```c
+typedef enum _sai_dash_flow_entry_bulk_get_session_mode_t
+
+{
+  SAI_DASH_FLOW_ENTRY_BULK_GET_SESSION_MODE_GRPC,
+
+  SAI_DASH_FLOW_ENTRY_BULK_GET_SESSION_MODE_EVENT,
+  
+  SAI_DASH_FLOW_ENTRY_BULK_GET_SESSION_MODE_EVENT_WITHOUT_FLOW_STATE,
+
+} sai_dash_flow_entry_bulk_get_session_mode_t;
+```
+
+#### Bulk Get Session Event Notifiations
+
+| Attribute name                                     | Type                                              | Description                                                  |
+| -------------------------------------------------- | ------------------------------------------------- | ------------------------------------------------------------ |
+| SAI_SWITCH_ATTR_FLOW_BULK_GET_SESSION_EVENT_NOTIFY | `sai_flow_bulk_get_session_event_notification_fn` | The callback function for receiving events on flow bulk get session event notification. |
+
+```c
+/**
+ * @brief bulk flow get event type
+ */
+typedef enum _sai_flow_bulk_get_session_event_t
+{
+    SAI_FLOW_BULK_GET_SESSION_FINISHED,
+
+    SAI_FLOW_BULK_GET_SESSION_FLOW_ENTRY,
+
+} sai_flow_bulk_get_session_event_t;
+
+/**
+ * @brief Notification data format received from SAI HA set callback
+ *
+ * @count attr[attr_count]
+ */
+typedef struct _sai_flow_bulk_get_session_event_data_t
+{
+    sai_flow_bulk_get_session_event_t event_type;
+
+    sai_object_id_t flow_bulk_session_id;
+  
+    sai_flow_entry_t *flow_entry;
+  
+    uint32_t attr_count; 
+  
+    sai_attribute_t *attr_list; 
+
+} sai_flow_bulk_get_session_event_data_t;
+
+/**
+ * @brief dash flow get bulk session notification
+ *
+ * Passed as a parameter into sai_initialize_switch()
+ *
+ * @count data[count]
+ *
+ * @param[in] count Number of notifications
+ * @param[in] data Array of flow bulk get session events
+ */
+typedef void (*sai_flow_bulk_get_session_event_notification_fn)(
+        _In_ uint32_t count,
+        _In_ const sai_flow_bulk_get_session_event_data_t *flow_bulk_get_session_event_data);
+```
+
+### Protobuf-based flow programming
+
+In addition to the flow state attributes, the flow state can be represented using protobuf for high-efficiency. The attribute of the flow entry is SAI_FLOW_ENTRY_ATTR_FLOW_DATA_PB.
+
+Although the content of both attributes and protobuf may be identical, their applications differ. Attributes enable incremental updates to individual properties, whereas protobuf necessitates a complete update.
+
+```protobuf
+syntax = "proto3";
+
+message SaiDashFlowKey {
+  uint64 eni_mac = 1;
+  string src_addr = 2;
```

Comment:
a more explicit type for mac and ip might be better. we can probably follow what SAI type is doing.